openjdk8/src/main/docker/shutdown-wrapper.bash

```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+function java_shutdown_hook {
+  # send thread dump to a file
+  if is_true "${SHUTDOWN_LOGGING_THREAD_DUMP}"; then
+    export THREAD_DUMP_FILE=${THREAD_DUMP_FILE:-/app.shutdown.threads}
+    jcmd $PID Thread.print  >> $THREAD_DUMP_FILE 2>&1
+  fi
+
+  # send heap info to a file
+  if is_true "${SHUTDOWN_LOGGING_HEAP_INFO}"; then
+    export HEAP_INFO_FILE=${HEAP_INFO_FILE:-/app.shutdown.heap}
+    jcmd $PID GC.class_histogram >> $HEAP_INFO_FILE 2>&1
+  fi
+}
+
+# capture the TERM signal and first generate the thread dump and/or heap info
+trap 'java_shutdown_hook; kill -TERM $PID' TERM
+$@ &
+PID=$!
+wait $PID
+wait $PID
```

Comment:
There's some explanation [here](http://veithen.github.io/2014/11/16/sigterm-propagation.html).
As far as I understand, the first wait is for when the JVM receives the signal, and the second is for when it terminates.